### PR TITLE
Self-sufficient account ref-counting

### PR DESCRIPTION
**Pull request by gavofyork**
Introduces a new ref-counted account type next to providers and consumers, called self-sufficient references. These allow an account to exist, but do not in themselves allow consumers to exist (for that there must be at least one provider reference on the account).

This is useful for times where you know that an aspect of an account is enough to pay for its basic existence, but where you want to guarantee that the reference can be yanked (probably by someone other than the account holder) without concern of there being consumers which depend on it.

Depended on by #8220 

Requires audit.

Includes migration!

### CODE REVIEW

**Comment by gavofyork** [comment]
CC @xlc

File: frame/system/src/lib.rs [hunk 1]
```diff
@@ -1070,11 +1083,63 @@ impl<T: Config> Module<T> {
...
+			}
+		})
+	}
+
+	/// The number of outstanding provider references for the account `who`.
 	pub fn providers(who: &T::AccountId) -> RefCount {
 		Account::<T>::get(who).providers
 	}
 
+	/// The number of outstanding provider references for the account `who`.
```
Comment (jacogr) [inline frame/system/src/lib.rs:1132]:
```suggestion
	/// The number of outstanding sufficient references for the account `who`.
```

File: frame/system/src/lib.rs [hunk 2]
```diff
@@ -262,9 +262,9 @@ pub mod pallet {
 	#[pallet::hooks]
 	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
 		fn on_runtime_upgrade() -> frame_support::weights::Weight {
-			if !UpgradedToDualRefCount::<T>::get() {
-				UpgradedToDualRefCount::<T>::put(true);
-				migrations::migrate_to_dual_ref_count::<T>()
+			if !UpgradedToTripleRefCount::<T>::get() {
+				UpgradedToTripleRefCount::<T>::put(true);
+				migrations::migrate_to_triple_ref_count::<T>()
```
Comment (bkchr) [inline frame/system/src/lib.rs:267]: Can we not call the migration explicitly in the particular runtimes?

**Comment by xlc** [comment]
I think I understand the purpose of this but still think the whole account ref system is under utilized & under documented. Will be helpful to have some guides to explain how it should be used along with examples. https://github.com/paritytech/substrate/issues/5255

File: frame/system/src/lib.rs (hunk 2 above)
Comment (gavofyork) [reply frame/system/src/lib.rs:267]: sure - i'm just concerned about it being forgotten.

File: frame/system/src/lib.rs [hunk 3]
```diff
@@ -630,14 +630,25 @@ mod migrations {
 	#[allow(dead_code)]
 	pub fn migrate_all<T: Config>() -> frame_support::weights::Weight {
```
Comment (thiolliere) [inline frame/system/src/lib.rs:631]:
```suggestion
	/// Migrate from unique `u8` reference counting to triple `u32` reference counting.
	pub fn migrate_all<T: Config>() -> frame_support::weights::Weight {
```

File: frame/system/src/lib.rs [hunk 4]
```diff
@@ -630,14 +630,25 @@ mod migrations {
...
 		Account::<T>::translate::<(T::Index, u8, T::AccountData), _>(|_key, (nonce, rc, data)|
-			Some(AccountInfo { nonce, consumers: rc as RefCount, providers: 1, data })
+			Some(AccountInfo { nonce, consumers: rc as RefCount, providers: 1, sufficients: 0, data })
 		);
 		T::BlockWeights::get().max_block
 	}
 
+	#[allow(dead_code)]
 	pub fn migrate_to_dual_ref_count<T: Config>() -> frame_support::weights::Weight {
```
Comment (thiolliere) [inline frame/system/src/lib.rs:639]:
```suggestion
	/// Migrate from unique `u32` reference counting to triple `u32` reference counting.
	pub fn migrate_to_dual_ref_count<T: Config>() -> frame_support::weights::Weight {
```

File: frame/system/src/lib.rs [hunk 5]
```diff
@@ -630,14 +630,25 @@ mod migrations {
...
-		Account::<T>::translate::<(T::Index, RefCount, T::AccountData), _>(|_key, (nonce, rc, data)|
-			Some(AccountInfo { nonce, consumers: rc as RefCount, providers: 1, data })
+		Account::<T>::translate::<(T::Index, RefCount, T::AccountData), _>(|_key, (nonce, consumers, data)|
+			Some(AccountInfo { nonce, consumers, providers: 1, sufficients: 0, data })
+		);
+		T::BlockWeights::get().max_block
+	}
+
+	/// Must be used.
+	pub fn migrate_to_triple_ref_count<T: Config>() -> frame_support::weights::Weight {
```
Comment (thiolliere) [inline frame/system/src/lib.rs:647]:
```suggestion
	/// Migrate from dual `u32` reference counting to triple `u32` reference counting.
	pub fn migrate_to_triple_ref_count<T: Config>() -> frame_support::weights::Weight {
```

File: frame/system/src/lib.rs [hunk 6]
```diff
@@ -580,10 +580,10 @@ pub mod pallet {
 	#[pallet::storage]
 	pub(super) type UpgradedToU32RefCount<T: Config> = StorageValue<_, bool, ValueQuery>;
 
-	/// True if we have upgraded so that AccountInfo contains two types of `RefCount`. False
+	/// True if we have upgraded so that AccountInfo contains three types of `RefCount`. False
 	/// (default) if not.
 	#[pallet::storage]
-	pub(super) type UpgradedToDualRefCount<T: Config> = StorageValue<_, bool, ValueQuery>;
```
[the remark this answers is not in the document]
Comment (gavofyork) [reply frame/system/src/lib.rs:586]: Once the auto-migration is removed, yes.

File: frame/system/src/lib.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (gavofyork) [reply frame/system/src/lib.rs:631]: That's horrible code (structs whose name is a prefixed by two underscores is pure evil) and completely undocumented. I agree that there should be a way of avoiding the use of naked tuples, but we need a properly supported API for expressing historic migrations whose types no longer exist in the main codebase, not magic incantations and special boilerplate which is incomprehensible to all but experts in the underlying implementation.

Until the APIs are clear, then I would stick to code which is admittedly less than optimal but at least comprehensible and not reliant on exotic and undocument patterns and APIs.

[the remark this answers is not in the document]
Comment (thiolliere) [reply frame/system/src/lib.rs:646]: yes ok, struct with underscore like `__Candidates` should be renamed `CandidatesStoragePrefixes`. And it is a lot of boilerplate.

hopefully we can make use of const generic to define the prefix to use of the storage in a simpler manner than this struct on which we have to implement the trait StorageInstance.

About documentation these are just type alias and a trait implementation, but I agree the migration file lacks documentation and the boilerplate is too much.
https://substrate.dev/rustdocs/v3.0.0/frame_support/storage/types/struct.StorageValue.html
https://substrate.dev/rustdocs/v3.0.0/frame_support/traits/trait.StorageInstance.html

**Review by shawntabrizi — APPROVED** [review]
lgtm

File: frame/system/src/lib.rs [hunk 7]
```diff
@@ -1051,30 +1065,34 @@ impl<T: Config> Module<T> {
...
-							"Logic error: Unexpected underflow in reducing provider",
-						);
-						Ok(DecRefStatus::Reaped)
-					},
-					(1, 0) => {
+				if account.providers == 0 {
+					// Logic error - cannot decrement beyond zero.
+					log::error!(
+						target: "runtime::system",
+			      "Logic error: Unexpected underflow in reducing provider",
```
Comment (kianenigma) [inline frame/system/src/lib.rs:1078]:
```suggestion
						"Logic error: Unexpected underflow in reducing provider",
```

Do I see spaces here? :P

**Review by kianenigma — COMMENTED** [review]
Logic within the PR lgtm, haven't looked at the linked PR yet. I do agree with @xlc that a guide system refernces would be needed. I also have some doubts about why and how we increment them in some pallets and not in others (e.g. https://github.com/paritytech/substrate/issues/7305)

**Review by thiolliere — APPROVED** [review]

File: frame/system/src/lib.rs (hunk 7 above)
Comment (gavofyork) [reply frame/system/src/lib.rs:1078]: Github's horrible merge tool :)

**Pull request merged**